// checkly/dist/constructs/check-group-ref.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckGroupRef = void 0;
const check_group_v1_1 = require("./check-group-v1");
const construct_1 = require("./construct");
const project_1 = require("./project");
/**
 * Creates a reference to an existing Check Group (both v1 and v2).
 *
 * References link existing resources to a project without managing them.
 */
class CheckGroupRef extends construct_1.Construct {
    constructor(logicalId, physicalId) {
        super(check_group_v1_1.CheckGroupV1.__checklyType, logicalId, physicalId, false);
        project_1.Session.registerConstruct(this);
    }
    describe() {
        return `CheckGroupRef:${this.logicalId}`;
    }
    getCheckDefaults() {
        // The only value CheckGroup.getCheckDefaults() returns is frequency,
        // which exists purely for convenience, and only at CLI-level. It never
        // gets sent to the backend. If references are being used, no access to
        // this feature is required, and it can be ignored entirely.
        return {};
    }
    getBrowserCheckDefaults() {
        // See the comment for getCheckDefaults(), the same applies here.
        return {};
    }
    getMultiStepCheckDefaults() {
        // See the comment for getCheckDefaults(), the same applies here.
        return {};
    }
    synthesize() {
        return null;
    }
}
exports.CheckGroupRef = CheckGroupRef;
