// checkly/dist/constructs/private-location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateLocation = exports.PrivateLocationRef = void 0;
const construct_1 = require("./construct");
const construct_diagnostics_1 = require("./construct-diagnostics");
const project_1 = require("./project");
/**
 * Creates a reference to an existing Private Location.
 *
 * References link existing resources to a project without managing them.
 */
class PrivateLocationRef extends construct_1.Construct {
    constructor(logicalId, physicalId) {
        super(PrivateLocation.__checklyType, logicalId, physicalId, false);
        project_1.Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocationRef:${this.logicalId}`;
    }
    synthesize() {
        return null;
    }
}
exports.PrivateLocationRef = PrivateLocationRef;
const RE_SLUG = /^((?!((us(-gov)?|ap|ca|cn|eu|sa|af|me)-(central|(north|south)?(east|west)?)-\d+))[a-zA-Z0-9-]{1,30})$/;
/**
 * Creates a Private Location
 *
 * @remarks
 *
 * This class make use of the Private Location endpoints.
 *
 * {@link https://www.checklyhq.com/docs/cli/constructs-reference/#privatelocation Read more in the docs}
 */
class PrivateLocation extends construct_1.Construct {
    name;
    slugName;
    icon;
    proxyUrl;
    static __checklyType = 'private-location';
    /**
     * Constructs the Private Location instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props private location configuration properties
     */
    constructor(logicalId, props) {
        super(PrivateLocation.__checklyType, logicalId);
        this.name = props.name;
        this.slugName = props.slugName;
        this.icon = props.icon;
        this.proxyUrl = props.proxyUrl;
        project_1.Session.registerConstruct(this);
    }
    describe() {
        return `PrivateLocation:${this.logicalId}`;
    }
    async validate(diagnostics) {
        await super.validate(diagnostics);
        if (!RE_SLUG.test(this.slugName)) {
            diagnostics.add(new construct_diagnostics_1.InvalidPropertyValueDiagnostic('slugName', new Error(`Value must not equal any AWS location.`)));
        }
    }
    static fromId(id) {
        return new PrivateLocationRef(`private-location-${id}`, id);
    }
    allowInChecklyConfig() {
        return true;
    }
    synthesize() {
        return {
            name: this.name,
            slugName: this.slugName,
            icon: this.icon,
            proxyUrl: this.proxyUrl,
        };
    }
}
exports.PrivateLocation = PrivateLocation;
