// checkly/dist/commands/baseCommand.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCommand = void 0;
const axios_1 = __importDefault(require("axios"));
const prompts_1 = __importDefault(require("prompts"));
const core_1 = require("@oclif/core");
const api_1 = require("../rest/api");
const command_style_1 = require("../helpers/command-style");
const resolver_1 = require("../services/check-parser/package-files/resolver");
class BaseCommand extends core_1.Command {
    static coreCommand = false;
    static hidden = true;
    fancy = true;
    style = new command_style_1.CommandStyle(this);
    #packageJsonLoader;
    async loadPackageJsonOfSelf() {
        if (!this.#packageJsonLoader) {
            const resolver = new resolver_1.PackageFilesResolver();
            this.#packageJsonLoader = resolver.loadPackageJsonFile(__filename);
        }
        return await this.#packageJsonLoader;
    }
    async checkEngineCompatibility() {
        const nodeVersion = process.versions.node;
        if (nodeVersion === undefined) {
            // Do nothing if Node version is not present. Note that Bun and Deno
            // do set a value for it though.
            return;
        }
        const packageJson = await this.loadPackageJsonOfSelf();
        if (packageJson === undefined) {
            // Do nothing if there's no package.json.
            return;
        }
        const { incompatible, requirements, } = packageJson.supportsEngine('node', nodeVersion);
        if (!incompatible) {
            // Do nothing if not incompatible.
            return;
        }
        this.style.longWarning(`Unsupported Node version`, `You are using Node v${nodeVersion}, which is not compatible with `
            + `the Checkly CLI. While you may continue to use the CLI, please be `
            + `advised that some functionality may not function correctly. Please `
            + `update to a newer version as soon as possible.`
            + `\n\n`
            + `We currently support the following versions:`
            + `\n\n`
            + `  ${requirements}`);
    }
    async init() {
        await this.checkEngineCompatibility();
        let version = process.env.CHECKLY_CLI_VERSION ?? this.config.version;
        // use latest version from NPM if it's running from the local environment or E2E
        if (version === '0.0.1-dev' || version?.startsWith('0.0.0')) {
            try {
                const { data: packageInformation } = await axios_1.default.get('https://registry.npmjs.org/checkly/latest');
                this.log(`\nNotice: replacing version '${version}' with latest '${packageInformation.version}'. If you wish to test with a different version, please pass the CHECKLY_CLI_VERSION environment variable.\n`);
                version = packageInformation.version;
            }
            catch {
                // No-op
            }
        }
        api_1.api.defaults.headers['x-checkly-cli-version'] = version;
        // This overrides prompts answers/selections (used on E2E tests)
        if (process.env.CHECKLY_E2E_PROMPTS_INJECTIONS) {
            try {
                const injections = JSON.parse(process.env.CHECKLY_E2E_PROMPTS_INJECTIONS);
                prompts_1.default.inject(injections);
            }
            catch {
                process.stderr.write('Error parsing CHECKLY_E2E_PROMPTS_INJECTIONS environment variable for injections.');
            }
        }
        if (process.env.CHECKLY_E2E_DISABLE_FANCY_OUTPUT) {
            this.fancy = false;
        }
        return super.init();
    }
    async run() {
        await this.exit(0);
    }
    catch(err) {
        // TODO: we can add Sentry here and log critical errors.
        return super.catch(err);
    }
}
exports.BaseCommand = BaseCommand;
